altair/utils/plugin_registry.py

```diff
@@ -1,15 +1,33 @@
 from __future__ import annotations
 
 from functools import partial
-from typing import Any, Generic, TypeVar, cast, Callable, TYPE_CHECKING
+from typing import Any, Generic, cast, Callable, TYPE_CHECKING
+from typing_extensions import TypeAliasType, TypeVar, TypeIs
 
 from importlib.metadata import entry_points
 
+from altair.utils.deprecation import deprecated_warn
+
 if TYPE_CHECKING:
     from types import TracebackType
 
+T = TypeVar("T")
+R = TypeVar("R")
+Plugin = TypeAliasType("Plugin", Callable[..., R], type_params=(R,))
+PluginT = TypeVar("PluginT", bound=Plugin[Any])
+IsPlugin = Callable[[object], TypeIs[Plugin[Any]]]
```

Comment:
As an example of how this could be used:

### alt.utils.theme.py
```py
from .plugin_registry import PluginRegistry
from typing import Callable
from typing_extensions import TypeAlias, TypeIs

ThemeType: TypeAlias = Callable[..., dict]

def is_theme_plugin(obj: Callable[..., Any]) -> TypeIs[ThemeType]:
    from inspect import signature
    from typing import get_origin

    sig = signature(obj)
    ret = sig.return_annotation
    return ret is dict or get_origin(ret) is dict

class ThemeRegistry(PluginRegistry[ThemeType, dict]):
    pass
```

### alt.vegalite.v5.theme.py
```py
from typing import Final
from ...utils.theme import ThemeRegistry
from altair.utils.plugin_registry import is_theme_plugin


ENTRY_POINT_GROUP: Final = "altair.vegalite.v5.theme"
themes = ThemeRegistry(entry_point_group=ENTRY_POINT_GROUP, plugin_type=is_theme_plugin)
```